rpcs3/Emu/SysCalls/Modules/cellGcmSys.cpp

```diff
@@ -1158,7 +1174,54 @@ s32 cellGcmCallback(vm::ptr<CellGcmContextData> context, u32 count)
 {
 	cellGcmSys.Log("cellGcmCallback(context_addr=0x%x, count=0x%x)", context.addr(), count);
 
-	if (1)
+	unsigned currentContext = 0;
+	for (const CellGcmContextData &ctx : contexts)
+	{
+		u32 curr = context->current;
+		if (curr <= ctx.end &&
+			curr >= ctx.begin)
+			break;
+		currentContext++;
+	}
+
+	if (currentContext == contexts.size() - 1)
+		currentContext = 0;
+	else
+		currentContext++;
+	GSLockCurrent gslock(GS_LOCK_WAIT_FLUSH);
+	auto& ctrl = vm::get_ref<CellGcmControl>(gcm_info.control_addr);
+	const std::chrono::time_point<std::chrono::system_clock> enterWait = std::chrono::system_clock::now();
+
+	while (true)
+	{
+		u32 getPos = ctrl.get.read_sync();
+		if (getPos < contexts[currentContext].begin || getPos > contexts[currentContext].end)
+			break;
+		std::chrono::time_point<std::chrono::system_clock> waitPoint = std::chrono::system_clock::now();
+		long long elapsedTime = std::chrono::duration_cast<std::chrono::seconds>(waitPoint - enterWait).count();
+		if (elapsedTime > 0)
+			cellGcmSys.Error("Has wait for more than a second for command buffer to be released by RSX");
+
+		std::this_thread::yield();
+	}
+
+	u32 address = contexts[currentContext].begin;
+	const u32 upper = offsetTable.ioAddress[address >> 20]; // 12 bits
+	assert(upper != 0xFFFF);
+	const u32 offset = (upper << 20) | (address & 0xFFFFF);
+
+	// TODO : it should be context->current but for some reasons it's not working
```

Comment:
I checkec original gcm callback functionality. It must use `current`, but writing `ctrl.put` must be the very first operation of `cellGcmCallback()`.